File: hw3/py/shape/glshape.py
"""
STOP. You should not modify this file unless you KNOW what you are doing.
"""

import copy

from OpenGL.GL import *
import glm

from util import Shader



class GLShape: 
    """
    Generic Shape object that manages the OpenGL context for a shape.
    All shapes directly interacting with the OpenGL context
    should public-inherit this class.
    """
    def __init__(self, 
                 shader: Shader, 
                 model: glm.mat4 = glm.mat4(1.0)):
        self.shader: Shader = shader
        self.vao: int = glGenVertexArrays(1)
        self.vbo: int = glGenBuffers(1)
        self.model: glm.mat4 = copy.deepcopy(model)
    
    def __del__(self):
        glDeleteVertexArrays(1, (self.vao,))
        self.vao = 0
        glDeleteBuffers(1, (self.vbo,))
        self.vbo = 0

File: hw3/cpp/src/shape/GLShape.cpp
/// STOP. You should not modify this file unless you KNOW what you are doing.

#include "shape/GLShape.h"


GLShape::~GLShape() noexcept
{
    glDeleteVertexArrays(1, &vao);
    vao = 0U;

    glDeleteBuffers(1, &vbo);
    vbo = 0U;
}


GLShape::GLShape(Shader * pShader, const glm::mat4 & model) : pShader(pShader), model(model)
{
    glGenVertexArrays(1, &vao);
    glGenBuffers(1, &vbo);
}


GLShape::GLShape(GLShape && rhs) noexcept
{
    *this = std::move(rhs);
}


GLShape & GLShape::operator=(GLShape && rhs) noexcept
{
    if (this == &rhs)
    {
        return *this;
    }

    pShader = rhs.pShader;
    rhs.pShader = nullptr;

    vao = rhs.vao;
    rhs.vao = 0U;

    vbo = rhs.vbo;
    rhs.vbo = 0U;

    model = rhs.model;

    return *this;
}
